ethik/base_explainer.py

```diff
@@ -8,15 +8,48 @@
 import numpy as np
 import pandas as pd
 import plotly.graph_objs as go
+from scipy import optimize
 from scipy import special
 from tqdm import tqdm
 
-from .utils import decimal_range, join_with_overlap, to_pandas, yield_masks
+from .utils import join_with_overlap, to_pandas, yield_masks
 from .warnings import ConvergenceWarning
 
 __all__ = ["BaseExplainer"]
 
 
+class ConvergenceSuccess(Exception):
+    """Custom warning to capture convergence success.
+
+    This is necessary to monkey-patch scipy's minimize function in order to manually stop the
+    minimization process.
+
+    """
+
+
+class F:
+    def __init__(self, x, target_mean, tol=3):
+        self.x = x
+        self.target_mean = target_mean
+        self.tol = tol
+
+    def __call__(self, ksi):
+        """Returns the loss and the gradient for a particular ksi value."""
+
+        self.ksi = ksi[0]
```

Comment:
Why do you need to store `ksi` in an attribute?

Reply:
I don't see any other way to retrieve the result... When the `ConvergenceSuccess` exception is raised, the `res` assignment isn't ackowledged, and thus we can't get access to the result. I know it's ugly. Someone smarter than me will probably see a better way! In the future, once `scipy` uses a class-based optimizers, this will most certainly be much cleaner.

Reply:
https://stackoverflow.com/a/12554345?

Reply:
That's a good idea... Give me a few minutes.

Reply:
Do you think you'll do it today? Otherwise, I can.

Reply:
I did 4 days ago! I just forgot to push it, it's done now. Sorry.